server/lib/timeline.ts

```diff
@@ -0,0 +1,222 @@
+      if (count > FEED_LIMIT) {
+        await redis.zRemRangeByRank(cacheKey, 0, count - FEED_LIMIT - 1);
+      }
+    }
+  }
+  stop();
+};
+
+export const getCollectiveFeed = async ({
+  collective,
+  dateTo,
+  limit = 20,
+  classes,
+}: {
+  collective: Collective;
+  dateTo: Date;
+  limit: number;
+  classes: ActivityClasses[];
+}) => {
+  const redis = await createRedisClient();
+  // If we don't have a redis client, we can't cache the timeline using sorted sets
+  if (!redis) {
+    debug('Redis is not configured, skipping cached timeline');
+    const stop = utils.stopwatch('timeline.readPage.noCache');
+    const where = await makeTimelineQuery(collective, classes);
+    if (dateTo) {
+      where['createdAt'] = { [Op.lt]: dateTo };
+    }
+
+    const activities = await models.Activity.findAll({
+      where,
+      order,
+      limit,
+    });
+    stop();
+    return activities;
+  }
+
+  // Check if timeline cache exists
+  const cacheKey = `timeline-${collective.slug}`;
+  const cacheExists = await redis.exists(cacheKey);
+  if (!cacheExists) {
+    // If we don't have a cache, generate it asynchronously
+    createOrUpdateFeed(collective);
+    return null;
+  }
+
+  const wantedTypes = flatten(classes.map(c => ActivitiesPerClass[c]));
+  const ids = [];
+  let offset = 0;
+
+  // This is a thunk responsible for paginating until we have enough results
+  const fetchMore = dateTo
+    ? // When fetching since a specific ID, we need to paginate in reverse by score (ID)
+      () =>
+        redis.zRange(cacheKey, `(${toInteger(dateTo.getTime() / 1000)}`, '-inf', {
+          BY: 'SCORE',
+          REV: true,
+          LIMIT: { count: limit, offset },
+        })
+    : // When fetching the latest activities, we can paginate in reverse by rank
+      () => redis.zRange(cacheKey, offset, offset + limit, { REV: true });
+
+  let cached = await fetchMore();
+
+  // If we're not paginating, first update cache
+  if (!dateTo) {
+    const [latest] = await redis.zRange(cacheKey, -1, -1);
+    const activity = JSON.parse(latest) as SerializedActivity;
+    debug(`Updating timeline for ${collective.slug} from id ${activity.id}`);
+    await createOrUpdateFeed(collective, activity.id);
+  }
+
+  const stop = utils.stopwatch('timeline.readPage.cached');
+  while (cached.length > 0) {
+    cached
+      .map(v => JSON.parse(v) as SerializedActivity)
+      // Filter out unwanted types based on the classes the user requested
+      .filter(({ type }) => wantedTypes.includes(type))
+      // In the case we fetch more than we need, we slice the array to the limit
+      .slice(0, limit - ids.length)
+      .forEach(({ id }) => ids.push(id));
+    if (ids.length >= limit) {
+      break;
+    } else {
+      offset += limit;
+      cached = await fetchMore();
+    }
+  }
+
+  // Return the actual activities from the database
+  const activities = await models.Activity.findAll({ where: { id: ids }, order });
+  stop();
```

Comment:
I would argue for renaming, otherwise you have to go up the code to understand what this function is

```suggestion
  stopWatch();
```